pkg/queue/sharedmain/main.go

```diff
@@ -304,18 +306,19 @@ func Main(opts ...Option) error {
 		}
 		logger.Info("Received TERM signal, attempting to gracefully shutdown servers.")
 		logger.Infof("Sleeping %v to allow K8s propagation of non-ready state", drainSleepDuration)
-		drainer.Drain()
-
-		// Removing the main server from the shutdown logic as we've already shut it down.
-		delete(httpServers, "main")
-
-		for serverName, srv := range httpServers {
-			logger.Info("Shutting down server: ", serverName)
-			if err := srv.Shutdown(context.Background()); err != nil {
-				logger.Errorw("Failed to shutdown server", zap.String("server", serverName), zap.Error(err))
+		drain(drainer, drainerTLS)
+
+		// Shutdown() to all services, wait no more than 600 sec for graceful termination
```

Comment:
Where is the 600 sec coming from? Do we have defined that somewhere (spec/docs)?

Reply:
No, it is not defined - do we need to define it? 

This is a fail-safe mechanism to ensure that we do terminate at some point after we received a KILL signal. Without it, we may hang on indefinitely due to any of our services not being able to terminate since it has an open connection.

Reply:
I will make that a const. Not very nice the way it is now :)

Reply:
Would it make sense to use something like the pod's termination grace period here (https://github.com/knative/serving/blob/0cf0da84f089a0af56761b6971a5d9a493949848/pkg/reconciler/revision/resources/deploy.go#L293)

Reply:
ok, so let's capture the status here first:

If we receive a PreStop:
- Immediately start draining the user container until QuietPeriod has elapsed since the last request
   (we have a fix that if we do not have a Kill Signal longer than one second later we will stop draining)

Once we receive a kill signal:
-  Immediately start draining the user container  until QuietPeriod (30s) has elapsed since the last request


Once we receive a kill signal and QuietPeriod (30s) has elapsed since the last request:
- We start shutting down the services, actually by that time all services should be idle - no? Maybe not all services but the main service at least?
- When we shutdown the services - since it is graceful, if there is an ongoing connection, we need to say for how long we wait (maybe this can be short since we already waited for 30s and all should be clean. But not specifying any value may hold us forever from terminating - come to think of it this probably can be few seconds.

- The TerminationGracePeriodSeconds seem to be intended as an overall measure for service pods - so maybe not be suitable for the time we wait in shutdown after we drained..

Reply:
/hold
After decision, need to fix the `ShutdownLimit` as it is no 5 minutes...